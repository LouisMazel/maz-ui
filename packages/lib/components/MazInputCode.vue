<template>
  <fieldset
    class="m-input-code"
    :class="[size ? `--${size}` : undefined, props.class]"
    :disabled="disabled"
    :style="style"
  >
    <div v-for="item in codeLength" :key="item" class="input-wrapper" :class="borderColorState">
      <!-- eslint-disable-next-line vuejs-accessibility/form-control-has-label -->
      <input
        :ref="
          (el) => {
            inputList[item - 1] = el as HTMLInputElement
          }
        "
        type="text"
        minlength="1"
        maxlength="1"
        :inputmode="acceptAlpha ? 'text' : 'numeric'"
        :pattern="acceptAlpha ? '[a-zA-Z0-9]{1}' : '[0-9]{1}'"
        autocomplete="do-not-autofill"
        :required="required"
        v-bind="$attrs"
        :value="inputValues.get(item)"
        @input="handleNewValue($event, item)"
        @keydown="handleKeydown($event, item)"
        @paste="setValueOnPaste"
      />
    </div>
  </fieldset>
</template>

<script lang="ts" setup>
  import { computed, ref, watch, type HTMLAttributes } from 'vue'
  import type { Color } from './types'

  type Size = 'xs' | 'sm' | 'md' | 'lg' | 'xl'

  export type { Color, Size }

  defineOptions({
    inheritAttrs: false,
  })

  const props = withDefaults(
    defineProps<{
      /** The style of the component. */
      style?: HTMLAttributes['style']

      /** The class of the component. */
      class?: HTMLAttributes['class']

      /** The value of the component (v-model). */
      modelValue?: string | number

      /** The length of the code. */
      codeLength?: number

      /** The type of the input field. */
      type?: 'text' | 'number'

      /** Whether to accept alpha characters. */
      acceptAlpha?: boolean

      /** Whether the input is required. */
      required?: boolean

      /** Whether the input is disabled. */
      disabled?: boolean

      /** Whether there is an error with the input. */
      error?: boolean

      /** Whether the input is successful. */
      success?: boolean

      /** Whether there is a warning with the input. */
      warning?: boolean

      /** The size of the component. */
      size?: Size

      /** The color of the component. */
      color?: Color
    }>(),
    {
      style: undefined,
      class: undefined,
      modelValue: undefined,
      codeLength: 4,
      type: 'text',
      acceptAlpha: false,
      required: false,
      disabled: false,
      error: false,
      success: false,
      warning: false,
      size: 'md',
      color: 'primary',
    },
  )

  const emits = defineEmits<{
    /**
     * Update the model value.
     * @param value The new value of the model.
     */
    (event: 'update:model-value', value?: string | number): void
    /**
     * Emitted when all inputs are set.
     */
    (event: 'completed'): void
  }>()

  const inputList = ref<HTMLInputElement[]>([])
  const localMap = ref<Map<number, string | undefined>>(new Map())

  watch(
    () => props.modelValue,
    (value, oldValue) => {
      if (value !== oldValue) {
        localMap.value = getMapValues()
      }
    },
    { immediate: true },
  )

  const inputValues = computed({
    get: () => localMap.value,
    set: (value) => {
      const emittedValue = getEmittedValue(value)
      emits('update:model-value', emittedValue)

      if (emittedValue?.toString().length === props.codeLength) {
        emits('completed')
      }
    },
  })

  function getMapValues(value = props.modelValue) {
    const map = new Map<number, string | undefined>()

    for (const item of Array.from({ length: props.codeLength }, (_, i) => i)) {
      if (value === undefined) {
        map.set(item + 1, undefined)
      } else {
        const values = [...value.toString()]

        map.set(item + 1, values[item])
      }
    }

    return map
  }

  function handleNewValue(event: Event, item: number) {
    const newValue = getValueSanitized((event.target as HTMLInputElement)?.value)

    inputValues.value = localMap.value.set(item, newValue)

    if (newValue) {
      focusAndSelectInputByIndex(item)
    }
  }

  function getEmittedValue(map: ReturnType<typeof getMapValues>) {
    const stringValue = [...map.values()].join('')

    return props.type === 'text'
      ? stringValue
      : stringValue && !Number.isNaN(Number(stringValue))
        ? Number(stringValue)
        : undefined
  }

  function getValueSanitized(value: string) {
    return (props.acceptAlpha ? value.match(/^[\w.]$/i) : value.match(/\d+/g))?.toString()
  }

  function handleKeydown(event: KeyboardEvent, inputIndex: number) {
    const currentInputValue = localMap.value.get(inputIndex)

    if (event.key === 'Backspace' && !currentInputValue) {
      const previousInputIndexToFocus = inputIndex - 1 < 0 ? 0 : inputIndex - 1

      inputValues.value = localMap.value.set(previousInputIndexToFocus, undefined)

      focusAndSelectInputByIndex(previousInputIndexToFocus - 1)
    }
  }

  function setValueOnPaste(event: ClipboardEvent) {
    const pasteData = event.clipboardData?.getData('text')

    if (!pasteData) {
      return
    }

    const indexInputsFromPastePlace = Array.from({ length: props.codeLength }, (_, i) => ({
      index: i + 1,
      value: ([...pasteData.toString()][i] ?? undefined) as string | undefined,
    }))

    for (const { index, value } of indexInputsFromPastePlace) {
      const sanitizedValue = value ? getValueSanitized(value) : undefined
      if (sanitizedValue) {
        localMap.value.set(index, sanitizedValue)
      }
    }

    setTimeout(() => {
      const lastInputWithoutValueIndex = getLastInputWithoutValueIndex()
      const lastIndex =
        lastInputWithoutValueIndex >= props.codeLength
          ? props.codeLength - 1
          : lastInputWithoutValueIndex
      focusAndSelectInputByIndex(lastIndex, false)
    }, 0)
  }

  function getLastInputWithoutValueIndex() {
    return [...localMap.value.values()].filter(Boolean).length
  }

  function focusAndSelectInputByIndex(index: number, selectValue = true) {
    setTimeout(() => {
      const input = inputList.value[index]

      if (index + 1 > props.codeLength || !input) {
        return
      }

      input.focus()
      if (selectValue) {
        input.select()
      }
    }, 0)
  }

  const borderColorState = computed(() => {
    if (props.error) return '!maz-border-danger'
    if (props.success) return '!maz-border-success'
    if (props.warning) return '!maz-border-warning'

    return ''
  })

  const borderColor = computed(() => {
    return `var(--maz-color-${props.color})`
  })
</script>

<style lang="postcss" scoped>
  .m-input-code {
    @apply maz-flex maz-gap-[1em];

    &.--mini {
      @apply maz-text-[0.625rem];

      line-height: 1rem;
    }

    &.--xs {
      @apply maz-text-xs;
    }

    &.--sm {
      @apply maz-text-sm;
    }

    &.--lg {
      @apply maz-text-lg;
    }

    &.--xl {
      @apply maz-text-xl;
    }

    .input-wrapper {
      @apply maz-relative maz-h-[4em] maz-w-[4em] maz-overflow-hidden maz-rounded maz-border maz-border-solid maz-border-gray-200 maz-transition-colors maz-duration-200 maz-ease-in-out dark:maz-border-color-lighter dark:maz-bg-color-light;

      &:focus-within {
        @apply maz-outline-none;

        border-color: v-bind('borderColor');
      }

      input {
        @apply maz-h-full maz-w-full maz-bg-transparent maz-text-center maz-text-[1.5em] maz-outline-none;
      }

      &:has(input:disabled) {
        @apply maz-bg-color-lighter maz-text-muted;

        input {
          @apply maz-cursor-not-allowed maz-text-muted;
        }
      }
    }
  }
</style>
