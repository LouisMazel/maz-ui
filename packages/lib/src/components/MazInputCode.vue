<script lang="ts" setup generic="T extends string | number">
import type { HTMLAttributes } from 'vue'
import type { MazColor } from './types'
import { computed, ref, watch } from 'vue'

export type MazInputCodeSize = 'xs' | 'sm' | 'md' | 'lg' | 'xl'
export interface MazInputCodeProps<T = string | number> {
  /** The style of the component. */
  style?: HTMLAttributes['style']
  /** The class of the component. */
  class?: HTMLAttributes['class']
  /** The value of the component (v-model). */
  modelValue?: T
  /** The length of the code. */
  codeLength?: number
  /** The type of the input field. */
  type?: 'text' | 'number'
  /** Whether to accept alpha characters. */
  acceptAlpha?: boolean
  /** Whether the input is required. */
  required?: boolean
  /** Whether the input is disabled. */
  disabled?: boolean
  /** Whether there is an error with the input. */
  error?: boolean
  /** Whether the input is successful. */
  success?: boolean
  /** Whether there is a warning with the input. */
  warning?: boolean
  /** The size of the component. */
  size?: MazInputCodeSize
  /** The color of the component. */
  color?: MazColor
  /** The hint text to display below the input. */
  hint?: string
}

defineOptions({
  inheritAttrs: false,
})

const props = withDefaults(defineProps<MazInputCodeProps<T>>(), {
  style: undefined,
  class: undefined,
  modelValue: undefined,
  codeLength: 4,
  type: 'text',
  acceptAlpha: false,
  required: false,
  disabled: false,
  error: false,
  success: false,
  warning: false,
  size: 'md',
  color: 'primary',
})

const emits = defineEmits<{
  /**
   * Update the model value.
   * @param value The new value of the model.
   */
  'update:model-value': [value?: T]
  /**
   * Emitted when all inputs are set.
   */
  'completed': [value: void]
}>()

const inputList = ref<HTMLInputElement[]>([])
const localMap = ref<Map<number, string | undefined>>(new Map())

watch(
  () => props.modelValue,
  (value, oldValue) => {
    if (value !== oldValue) {
      localMap.value = getMapValues()
    }
  },
  { immediate: true },
)

const inputValues = computed({
  get: () => localMap.value,
  set: (value) => {
    const emittedValue = getEmittedValue(value)
    emits('update:model-value', emittedValue as T)

    if (emittedValue?.toString().length === props.codeLength) {
      emits('completed')
    }
  },
})

function getMapValues(value = props.modelValue) {
  const map = new Map<number, string | undefined>()

  for (const item of Array.from({ length: props.codeLength }, (_, i) => i)) {
    if (value === undefined) {
      map.set(item + 1, undefined)
    }
    else {
      const values = [...value.toString()]

      map.set(item + 1, values[item])
    }
  }

  return map
}

function handleNewValue(event: Event, item: number) {
  const newValue = getValueSanitized((event.target as HTMLInputElement)?.value)

  inputValues.value = localMap.value.set(item, newValue)

  if (newValue) {
    focusAndSelectInputByIndex(item)
  }
}

function getEmittedValue(map: ReturnType<typeof getMapValues>) {
  const stringValue = [...map.values()].join('')

  if (props.type === 'text') {
    return stringValue
  }

  return stringValue && !Number.isNaN(Number(stringValue))
    ? Number(stringValue)
    : undefined
}

function getValueSanitized(value: string) {
  return (props.acceptAlpha ? value.match(/^[\w.]$/) : value.match(/\d+/g))?.toString()
}

function handleKeydown(event: KeyboardEvent, inputIndex: number) {
  const currentInputValue = localMap.value.get(inputIndex)

  if (event.key === 'ArrowRight') {
    focusAndSelectInputByIndex(inputIndex)
  }

  if (event.key === 'ArrowLeft') {
    focusAndSelectInputByIndex(inputIndex - 2)
  }

  if (event.key === 'Backspace' && !currentInputValue) {
    const previousInputIndexToFocus = inputIndex - 1 < 0 ? 0 : inputIndex - 1

    inputValues.value = localMap.value.set(previousInputIndexToFocus, undefined)

    focusAndSelectInputByIndex(previousInputIndexToFocus - 1)
  }
}

function setValueOnPaste(event: ClipboardEvent) {
  const pasteData = event.clipboardData?.getData('text')

  if (!pasteData) {
    return
  }

  const indexInputsFromPastePlace = Array.from({ length: props.codeLength }, (_, i) => ({
    index: i + 1,
    value: ([...pasteData.toString()][i] ?? undefined) as string | undefined,
  }))

  for (const { index, value } of indexInputsFromPastePlace) {
    const sanitizedValue = value ? getValueSanitized(value) : undefined
    if (sanitizedValue) {
      localMap.value.set(index, sanitizedValue)
    }
  }

  setTimeout(() => {
    const lastInputWithoutValueIndex = getLastInputWithoutValueIndex()
    const lastIndex
        = lastInputWithoutValueIndex >= props.codeLength
          ? props.codeLength - 1
          : lastInputWithoutValueIndex
    focusAndSelectInputByIndex(lastIndex, false)
  }, 0)
}

function getLastInputWithoutValueIndex() {
  return [...localMap.value.values()].filter(Boolean).length
}

function focusAndSelectInputByIndex(index: number, selectValue = true) {
  setTimeout(() => {
    const input = inputList.value[index]

    if (index + 1 > props.codeLength || !input) {
      return
    }

    input.focus()
    if (selectValue) {
      selectInputByIndex(index)
    }
  }, 0)
}

function selectInputByIndex(index: number) {
  const input = inputList.value[index]

  if (index + 1 > props.codeLength || !input) {
    return
  }

  input.select()
}

const borderColorState = computed(() => {
  if (props.error)
    return '!maz-border-destructive'
  if (props.success)
    return '!maz-border-success'
  if (props.warning)
    return '!maz-border-warning'

  return ''
})
</script>

<template>
  <fieldset
    class="m-input-code m-reset-css"
    :class="[size ? `--${size}` : undefined, props.class]"
    :disabled="disabled"
    :style="[style, { '--input-border-color': `hsl(var(--maz-${props.color}))` }]"
  >
    <div class="m-input-code__wrapper">
      <div v-for="item in codeLength" :key="item" class="input-wrapper" :class="borderColorState">
        <!-- eslint-disable-next-line vuejs-accessibility/form-control-has-label -->
        <input
          :ref="
            (el) => {
              inputList[item - 1] = el as HTMLInputElement
            }
          "
          type="text"
          minlength="1"
          maxlength="1"
          :inputmode="acceptAlpha ? 'text' : 'numeric'"
          :pattern="acceptAlpha ? '[a-zA-Z0-9]{1}' : '[0-9]{1}'"
          autocomplete="do-not-autofill"
          :required="required"
          v-bind="$attrs"
          :value="inputValues.get(item)"
          @input="handleNewValue($event, item)"
          @keydown="handleKeydown($event, item)"
          @click="selectInputByIndex(item - 1)"
          @paste="setValueOnPaste"
        >
      </div>
    </div>
    <span
      class="m-input-code__hint" :class="{
        '--error': error,
        '--success': success,
        '--warning': warning,
      }"
    >{{ hint }}</span>
  </fieldset>
</template>

<style lang="postcss" scoped>
  .m-input-code {
  @apply maz-inline-flex maz-flex-col maz-gap-[0.5em] maz-align-top;

  &.--mini {
    @apply maz-text-[0.625rem];

    line-height: 1rem;
  }

  &.--xs {
    @apply maz-text-xs;
  }

  &.--sm {
    @apply maz-text-sm;
  }

  &.--lg {
    @apply maz-text-lg;
  }

  &.--xl {
    @apply maz-text-xl;
  }

  &__wrapper {
    @apply maz-inline-flex maz-gap-[1em];
  }

  &__hint {
    @apply maz-text-sm maz-text-muted;

    &.--error {
      @apply maz-text-destructive-600;
    }

    &.--success {
      @apply maz-text-success-600;
    }

    &.--warning {
      @apply maz-text-warning-600;
    }
  }

  .input-wrapper {
    @apply maz-relative maz-h-[4em] maz-w-[4em] maz-overflow-hidden maz-rounded maz-border maz-border-solid maz-border-divider maz-transition-colors maz-duration-200 maz-ease-in-out dark:maz-bg-surface-400;

    &:focus-within {
      border-color: var(--input-border-color);
    }

    input {
      @apply maz-h-full maz-w-full maz-bg-transparent maz-text-center maz-text-[1.5em] maz-outline-none;
    }

    &:has(input:disabled) {
      @apply maz-bg-surface-300 maz-text-muted;

      input {
        @apply maz-cursor-not-allowed maz-text-muted;
      }
    }
  }
}
</style>
