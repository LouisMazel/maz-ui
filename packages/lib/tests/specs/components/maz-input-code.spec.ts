import { shallowMount, type VueWrapper } from '@vue/test-utils'
import MazInputCode from '@components/MazInputCode.vue'

describe('MazInputCode', () => {
  let wrapper: VueWrapper<InstanceType<typeof MazInputCode>>

  beforeEach(() => {
    wrapper = shallowMount(MazInputCode)
  })

  test('Should match with the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('renders the correct number of input fields', async () => {
    expect(wrapper.findAll('input')).toHaveLength(4)

    await wrapper.setProps({ codeLength: 6 })
    expect(wrapper.findAll('input')).toHaveLength(6)
  })

  test('emits completed event when code is complete', async () => {
    const inputFields = wrapper.findAll('input')
    await inputFields[0].setValue('1')
    await inputFields[1].setValue('2')
    await inputFields[2].setValue('3')
    await inputFields[3].setValue('4')

    expect(wrapper.emitted('completed')).toHaveLength(1)
  })

  test('sets initial input values correctly', async () => {
    await wrapper.setProps({
      modelValue: '1234',
    })

    const inputFields = wrapper.findAll('input')
    expect(inputFields[0].element.value).toBe('1')
    expect(inputFields[1].element.value).toBe('2')
    expect(inputFields[2].element.value).toBe('3')
    expect(inputFields[3].element.value).toBe('4')
  })

  test('allows only numeric values when acceptAlpha prop is false', async () => {
    const inputFields = wrapper.findAll('input')

    await inputFields[0].setValue('a')
    await inputFields[1].setValue('1')
    await inputFields[2].setValue('f')
    await inputFields[3].setValue('2')

    expect(wrapper.emitted('update:model-value')).toHaveLength(4)
    expect(wrapper.emitted('update:model-value')).toStrictEqual([[''], ['1'], ['1'], ['12']])
  })

  test('allows alphanumeric values when acceptAlpha prop is true', async () => {
    await wrapper.setProps({
      acceptAlpha: true,
    })

    const inputFields = wrapper.findAll('input')
    await inputFields[0].setValue('a')
    await inputFields[1].setValue('1')
    await inputFields[2].setValue('b')
    await inputFields[3].setValue('2')

    expect(inputFields[0].element.value).toBe('a')
    expect(inputFields[1].element.value).toBe('1')
    expect(inputFields[2].element.value).toBe('b')
    expect(inputFields[3].element.value).toBe('2')

    expect(wrapper.emitted('update:model-value')).toHaveLength(4)
    expect(wrapper.emitted('update:model-value')).toStrictEqual([['a'], ['a1'], ['a1b'], ['a1b2']])
  })
})
