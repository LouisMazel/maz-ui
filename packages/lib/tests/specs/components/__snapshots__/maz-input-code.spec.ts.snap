// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MazInputCode > Should match with the snapshot 1`] = `
"<fieldset data-v-757e2ef4=\\"\\" class=\\"m-input-code --md\\">
  <div data-v-757e2ef4=\\"\\" class=\\"input-wrapper\\">
    <!-- eslint-disable-next-line vuejs-accessibility/form-control-has-label --><input data-v-757e2ef4=\\"\\" type=\\"text\\" minlength=\\"1\\" maxlength=\\"1\\" inputmode=\\"numeric\\" pattern=\\"[0-9]{1}\\" autocomplete=\\"do-not-autofill\\">
  </div>
  <div data-v-757e2ef4=\\"\\" class=\\"input-wrapper\\">
    <!-- eslint-disable-next-line vuejs-accessibility/form-control-has-label --><input data-v-757e2ef4=\\"\\" type=\\"text\\" minlength=\\"1\\" maxlength=\\"1\\" inputmode=\\"numeric\\" pattern=\\"[0-9]{1}\\" autocomplete=\\"do-not-autofill\\">
  </div>
  <div data-v-757e2ef4=\\"\\" class=\\"input-wrapper\\">
    <!-- eslint-disable-next-line vuejs-accessibility/form-control-has-label --><input data-v-757e2ef4=\\"\\" type=\\"text\\" minlength=\\"1\\" maxlength=\\"1\\" inputmode=\\"numeric\\" pattern=\\"[0-9]{1}\\" autocomplete=\\"do-not-autofill\\">
  </div>
  <div data-v-757e2ef4=\\"\\" class=\\"input-wrapper\\">
    <!-- eslint-disable-next-line vuejs-accessibility/form-control-has-label --><input data-v-757e2ef4=\\"\\" type=\\"text\\" minlength=\\"1\\" maxlength=\\"1\\" inputmode=\\"numeric\\" pattern=\\"[0-9]{1}\\" autocomplete=\\"do-not-autofill\\">
  </div>
</fieldset>"
`;
